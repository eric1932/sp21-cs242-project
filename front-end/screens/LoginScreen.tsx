import React from "react";
import {StyleSheet, Text, TextInput, TouchableOpacity, View} from "react-native";
import {getToken, saveToken} from "../utils/Storage";
import {validateUser} from "../utils/API";
import {LoginProps, LoginScreenNavigationProp} from "../types";

function handleLogin(username: string, password: string, navigation: LoginScreenNavigationProp) {
  return async () => {
    let result: string | null = await validateUser(username, password)
    if (result) {
      await saveToken(result);
      navigation.replace('Root')
    }
  };
}

// ref: https://github.com/Alhydra/React-Native-Login-Screen-Tutorial
export default function LoginScreen({navigation}: LoginProps) {
  let [username, setUsername] = React.useState('')
  let [password, setPassword] = React.useState('')

  React.useEffect(() => {
    getToken().then(token => {
      if (token !== null) {
        navigation.replace('Root')
      }
    })
  })

  return (
    <View style={styles.container}>
      <Text style={styles.logo}>Daily Checkin</Text>
      <View style={styles.inputView}>
        <TextInput
          style={styles.inputText}
          placeholder="Username"
          placeholderTextColor="#003f5c"
          onChangeText={text => setUsername(text)}/>
      </View>
      <View style={styles.inputView}>
        <TextInput
          secureTextEntry
          style={styles.inputText}
          placeholder="Password"
          placeholderTextColor="#003f5c"
          onChangeText={text => setPassword(text)}/>
      </View>
      {/*<TouchableOpacity>*/}
      {/*  <Text style={styles.forgot}>Forgot Password?</Text>*/}
      {/*</TouchableOpacity>*/}
      <TouchableOpacity
        style={styles.loginBtn}
        onPress={handleLogin(username, password, navigation)}>
        <Text
          style={styles.loginText}>LOGIN</Text>
      </TouchableOpacity>
      <TouchableOpacity>
        <Text style={[styles.loginText, {paddingTop: 10}]}>Signup</Text>
      </TouchableOpacity>
    </View>
  )
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#003f5c',
    alignItems: 'center',
    justifyContent: 'center',
  },
  logo: {
    fontWeight: "bold",
    fontSize: 50,
    color: "#fb5b5a",
    marginBottom: 40
  },
  inputView: {
    width: "80%",
    backgroundColor: "#465881",
    borderRadius: 25,
    height: 50,
    marginBottom: 20,
    justifyContent: "center",
    padding: 20
  },
  inputText: {
    height: 50,
    color: "white"
  },
  forgot: {
    color: "white",
    fontSize: 11
  },
  loginBtn: {
    width: "80%",
    backgroundColor: "#fb5b5a",
    borderRadius: 25,
    height: 50,
    alignItems: "center",
    justifyContent: "center",
    marginTop: 40,
    marginBottom: 10
  },
  loginText: {
    color: "white",
  }
});