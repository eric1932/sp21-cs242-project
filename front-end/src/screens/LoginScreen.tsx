import React, {ReactElement} from "react";
import {StyleSheet, Text, TextInput, TouchableOpacity, View} from "react-native";
import {getToken, saveToken} from "../utils/Storage";
import {validateUser, validateUserToken} from "../utils/API";
import {LoginProps, LoginScreenNavigationProp} from "../types";

function handleLogin(useTokenLogin: boolean, username: string, password: string, token: string, navigation: LoginScreenNavigationProp) {
  // TODO fail message
  if (useTokenLogin) {
    return async () => {
      const result: boolean = await validateUserToken(token)
      if (result) {
        await saveToken(token);
        navigation.replace('Root')
      }
    };
  } else {
    return async () => {
      const result: string | null = await validateUser(username, password)
      if (result) {
        await saveToken(result);
        navigation.replace('Root')
      }
    };
  }
}

// ref: https://github.com/Alhydra/React-Native-Login-Screen-Tutorial
export default function LoginScreen({navigation}: LoginProps): ReactElement {
  const [useTokenLogin, setTokenLogin] = React.useState(false)
  const [username, setUsername] = React.useState('')
  const [password, setPassword] = React.useState('')
  const [token, setToken] = React.useState('')

  React.useEffect(() => {
    void getToken().then(token => {
      if (token !== null) {
        navigation.replace('Root')
      }
    })
  })

  return (
    <View style={styles.container}>
      <Text style={styles.logo}>Daily Checkin</Text>
      {useTokenLogin
        ? <View style={styles.inputView}>
          <TextInput
            style={styles.inputText}
            placeholder="Token"
            placeholderTextColor="#003f5c"
            onChangeText={text => setToken(text)}/>
        </View>
        : <><View style={styles.inputView}>
          <TextInput
            style={styles.inputText}
            placeholder="Username"
            placeholderTextColor="#003f5c"
            onChangeText={text => setUsername(text)}/>
        </View>
          <View style={styles.inputView}>
            <TextInput
              secureTextEntry
              style={styles.inputText}
              placeholder="Password"
              placeholderTextColor="#003f5c"
              onChangeText={text => setPassword(text)}/>
          </View></>}
      <TouchableOpacity onPress={() => setTokenLogin(!useTokenLogin)}>
        {useTokenLogin
          ? <Text style={styles.forgot}>use username and password</Text>
          : <Text style={styles.forgot}>or use token to login</Text>}
      </TouchableOpacity>
      <TouchableOpacity
        style={styles.loginBtn}
        onPress={handleLogin(useTokenLogin, username, password, token, navigation)}>
        <Text
          style={styles.loginText}>LOGIN</Text>
      </TouchableOpacity>
      <TouchableOpacity>
        <Text style={[styles.loginText, {paddingTop: 10}]}>Signup</Text>
      </TouchableOpacity>
    </View>
  )
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#003f5c',
    alignItems: 'center',
    justifyContent: 'center',
  },
  logo: {
    fontWeight: "bold",
    fontSize: 50,
    color: "#fb5b5a",
    marginBottom: 40
  },
  inputView: {
    width: "80%",
    backgroundColor: "#465881",
    borderRadius: 25,
    height: 50,
    marginBottom: 20,
    justifyContent: "center",
    padding: 20
  },
  inputText: {
    height: 50,
    color: "white"
  },
  forgot: {
    color: "white",
    fontSize: 11
  },
  loginBtn: {
    width: "80%",
    backgroundColor: "#fb5b5a",
    borderRadius: 25,
    height: 50,
    alignItems: "center",
    justifyContent: "center",
    marginTop: 40,
    marginBottom: 10
  },
  loginText: {
    color: "white",
  }
});