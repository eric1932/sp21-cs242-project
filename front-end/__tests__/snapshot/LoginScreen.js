import * as React from 'react';
import renderer from 'react-test-renderer';
import LoginScreen from "../../src/screens/LoginScreen";

const createTestProps = (props) => ({
  navigation: {
    navigate: jest.fn()
  },
  ...props
});

it('LoginScreen', () => {
  const tree = renderer.create(
    <LoginScreen {...createTestProps({})} />
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
