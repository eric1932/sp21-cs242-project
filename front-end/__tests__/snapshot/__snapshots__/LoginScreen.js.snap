// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginScreen 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#003f5c",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      Object {
        "color": "#fb5b5a",
        "fontSize": 50,
        "fontWeight": "bold",
        "marginBottom": 40,
      }
    }
  >
    Daily Checkin
  </Text>
  <View
    style={
      Object {
        "backgroundColor": "#465881",
        "borderRadius": 25,
        "height": 50,
        "justifyContent": "center",
        "marginBottom": 20,
        "padding": 20,
        "width": "80%",
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Username"
      placeholderTextColor="#003f5c"
      rejectResponderTermination={true}
      style={
        Object {
          "color": "white",
          "height": 50,
        }
      }
      underlineColorAndroid="transparent"
    />
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#465881",
        "borderRadius": 25,
        "height": 50,
        "justifyContent": "center",
        "marginBottom": 20,
        "padding": 20,
        "width": "80%",
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Password"
      placeholderTextColor="#003f5c"
      rejectResponderTermination={true}
      secureTextEntry={true}
      style={
        Object {
          "color": "white",
          "height": 50,
        }
      }
      underlineColorAndroid="transparent"
    />
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 11,
        }
      }
    >
      or use token to login
    </Text>
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#fb5b5a",
        "borderRadius": 25,
        "height": 50,
        "justifyContent": "center",
        "marginBottom": 10,
        "marginTop": 40,
        "opacity": 1,
        "width": "80%",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
        }
      }
    >
      LOGIN
    </Text>
  </View>
  <View
    accessible={true}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "white",
          },
          Object {
            "paddingTop": 10,
          },
        ]
      }
    >
      Signup
    </Text>
  </View>
</View>
`;
